Result support (#118)
Limit-Theory-Redux/ltheory: engine/lib/luajit-ffi-gen/readme.md

cargo expand -p luajit-ffi-gen --test basic_test
```

## Supported types

**luajit_ffi_gen** attribute supports different types as input parameters and return type in the impl methods.

Following table shows representation of Rust types in the generated code.

> | Rust type                            | C API               | Lua c_defs |

Reviewer: This feels more accurate to me:
```suggestion
| Rust type                            | extern "C" interface         | C type |
```

| ------------------------------------ | ------------------- | ---------- |
| Immutable reference (&T)             | &T                  | T const*   |
| Mutable reference (&mut T)           | &mut T              | T *        |
| Self (in return position)            | Box\<T>             | T *        |
| Basic and copy types (T)             | T                   | T          |
| String, str                          | *const libc::c_char | cstr       |
| Option\<T>                           | *mut T              | T *        |

> | Result\<T, E> (only return position) | T, panic on error   | T          |

Reviewer: I wonder whether it would be nice to generate a C struct like:

`Result<T, E>` becomes
```
struct Result_T_E {
  T ok;
  E error;
  bool is_ok;
};
```

then in the Lua FFI code, we transform that struct into either `ok, nil`, or return two values `nil, error`, something kind of like what Roblox does: https://roblox.github.io/lua-style-guide/#error-handling

Author: Good idea :+1: 
I was thinking about using lua_error() function to return an error back to Lua but unfortunately it will require passing lua_State everywhere.
Your approach is better.
This will require some work so we should do it in another PR.

Reviewer: Agreed

### (Mutable) references

In the generated C API references are used as is: **&** or **&mut**. Exceptions are only strings and options.

### Self

Self as input parameter is expected as reference (**&self**) or mutable reference (**&mut self**).

In the return position **Self** is boxed: Box\<T>.

### Basic and copy types

> Basic (bool, i32, u64, f32, etc.) and copy types (defined in **COPY_TYPES** map) are sent as is via C API.

Reviewer: ```suggestion
Basic (bool, i32, u64, f32, etc.) and copy types (defined in **COPY_TYPES** map) are sent as is via the C API.
```

### Strings

> To avoid additional copy C strings in method parameter position is converted unsafely to Rust string.

Reviewer: ```suggestion
To avoid additional copies, C strings in the method parameter position is converted unsafely to a Rust string.
```

> In return position Rust string is converted to CStr and copied to a static scoped buffer. Pointer to this buffer is returned via C API.

Reviewer: ```suggestion
When returning, Rust strings are converted to CStr and copied to a static scoped buffer. A pointer to this buffer is returned via the C API.
```

### User defined types (non copyable)

> User types in method parameters position are sent by reference or mutable reference. In the return position type is boxed to transfer ownership to caller (Lua).

Reviewer: ```suggestion
User types in method parameters position are sent by reference or mutable reference. When returning, the type is boxed to transfer ownership to the caller (Lua).
```

**TODO**: should we accept user types in parameters position boxed as well?

### Option

> Transferred as **\*mut T** and **None** interpreted as **NULL** pointer.

Reviewer: ```suggestion
Returned as a **\*mut T**, and **None** is nterpreted as **NULL** pointer.
```

### Result

Accepted only in the return position. Panics on error.

---
## Tip

If there are any compilation errors in the generated C API bindings, developer can copy **#[no_mangle]** code generated by **cargo expand** into the *.rs file and disable **#[luajit_ffi_gen]** attribute.

This way it will be much easier to spot the place of the problem.

## Optimization ideas

> If after converting engine to use **luajit_ffi_gen** attribute everywhere compilation time significantly increases, code generation can be optimized in 2 stages.

Reviewer: ```suggestion
If compilation time significantly increases after utilising the **luajit_ffi_gen** attribute, code generation can be optimized in 2 stages:
```

### Stage 1. Regenerate Lua FFI only if Rust code changes

> To avoid unnecessary Lua FFI files regeneration we can calculate hash of the [ImplInfo] structure (actually only name and methods fields) and store it in a file. So before generating Lua FFI file we can check if it's hash changed and do regeneration only if it does.

Reviewer: ```suggestion
To avoid unnecessary Lua FFI files regeneration we can calculate the hash of the [ImplInfo] structure and store it in a file. So before generating the Lua FFI code, we can check to see if it's hash changed first before regenerating.
```

> File with a hash can be stored either in **target** folder subfolder or in a git. For the former improvement will be visible only for the incremental build. The latter will help CI as well.

Reviewer: ```suggestion
The hash file can be stored either in a subfolder of the **target** directory, or in git. In the former case, improvement will be visible only during an incremental build, but the latter will help during CI as well.
```

### Stage 2. Regenerate C API bindings only if Rust code changes

> If optimization from the stage 1 is not enough similar approach can be applied to the generated C API code.

Reviewer: ```suggestion
If the optimization steps from the stage 1 is not enough, a similar approach can be applied to the generated C API code.
```

Reviewer 2: > If the optimization steps from stage 1 are not enough, a similar approach can be applied to the generated C API code.

> In this case it should not be put in the same file where **luajit_ffi_gen** was added but in a separate one.
> 
> Here we also do regeneration of the C API file only if hash of the [ImplInfo] structure was changed.

Reviewer: ```suggestion
In this case it should be placed in a different location to the hash of the Lua FFI code. 

Here we also do regeneration of the C API file only if the hash of the [ImplInfo] structure was changed.
```
